### DISCUSSION TITLE
JobDataMap stores empty string for null value

### DISCUSSION THREAD

**Discussion by asherber** [quartznet/quartznet#1582] [Q&A] [answered]
I have an Ado scheduler created with a job store with the JSON serializer and `UseProperties=true`, talking to a Sql Server database. I create a `JobDataMap` with `["foo"] = null` and schedule a job with that job data. Because of [this line](https://github.com/quartznet/quartznet/blob/ec044e9bfc67006c2a780a8817a9e175c079d8b2/src/Quartz/Impl/AdoJobStore/StdAdoDelegate.cs#L548), that null gets serialized and stored in the db as an empty string. 

Within my job class, `context.MergedDataMap.GetDateTimeValue("foo")` now throws an exception because the map has an empty string instead of the null I put in there.

I was going to file a bug report about this, but turning the null into an empty string during serialization seems to be a conscious choice, so I wanted first to ask why this was done.

**Comment by lahma** [comment, marked as answer]
I think the straight answer to this question is that this was ported 1:1 from Java version. 

But the thing is, that Java version does not have equivalent of `GetDateTimeValue`. And because the API hasn't been designed to return nullable types, this will of course fail. However, it should not be hard to write your own extension methods against `JobDataMap` like `TryGetDateTimeValue`. 

Even better, I think if you'd like to create a PR to cover such `Try*` alternatives for the methods it would be greatly appreciated!

**Reply by asherber** [reply]
Thanks, appreciate the answer. I think this behavior should be called out in the docs, that when `UseProperties==true`, any null value will get converted to `String.Empty`. (I also wish this were not the case -- I think if I store a null it should remain as a null -- but I understand that this would be a breaking change and would diverge from the Java functionality.)

I'd be happy to put together a PR for `Try*` functionality.

**Reply by lahma** [reply]
This is something hard to change at this point, as you also have concluded. It's not bad to diverge from Java version per se, but behavior change might be problematic also on .NET side (even if it would make things more logical). 

I'd gladly review such a `Try`-overload PR. There are many methods to cover, but should be quite idiomatic. If you kindly go this route, please make sure to target `v3.x` branch to make sure we add them to servicing release.

**Reply by asherber** [reply]
@lahma I can target `V3.x` with the PR -- but where would you like me to branch from? Also `V3.x`?

**Reply by lahma** [reply]
Yes, that would make the PR manageable. Otherwise it would bring new changes from main too.
